lib/src/group_changes.dart

```diff
@@ -0,0 +1,307 @@
+
+Future<Map<String, dynamic>> commitInformation() async {
+  final client = HttpClient();
+  final request = await client.getUrl(Uri.parse(
+      "https://dart.googlesource.com/sdk/+log/master?n=400&format=JSON"));
+  final response = await request.close();
+  return (await response
+      .transform(utf8.decoder)
+      .transform(LineSplitter())
+      .skip(1)
+      .transform(json.decoder)
+      .single) as Map<String, dynamic>;
+}
+
+/// Group all changes by test name, change in result (all three of
+/// previous result, new result, expected result in one key).
+/// Dynamically create subgroups from this sorted group by greedily
+/// taking all changes whose blamelist intersects with the blamelist
+/// that starts last, taking them out to form a subgroup, and repeating.
+/// For each subgroup, take the configurations of all the changes, and make
+/// this set of configurations a key (by sorting and toString).
+/// Store a data object summarizing this subgroup, with the test name,
+/// change in result, set of configurations,
+/// and blamelist intersection and union of the changes in this subgroup.
+/// The object is stored in a multi-level map keyed by blamelist intersection
+/// last commit, blamelist intersection first commit, and set of configurations.
+Map<int, Map<int, Map<String, List<ByConfigSetData>>>> computePageData(
+    List<dynamic> changes, List<String> hashes) {
+  final Map<String, int> hashIndex = Map.fromEntries(
+      Iterable.generate(hashes.length, (i) => MapEntry(hashes[i], i)));
+
+  final resultsForTestAndChange =
+      Map<String, Map<String, List<Map<String, dynamic>>>>();
+
+  for (final Map<String, dynamic> change in changes) {
+    String key;
+    if ('true' == change['matches'] ||
+        change['matches'] is bool && change['matches']) {
+      key =
+          "${change['previous_result']} -> <strong>${change['result']} &#x2714;</strong>";
+    } else {
+      key =
+          "${change['previous_result']} -> <strong>${change['result']}</strong>  (expected ${change['expected']})";
+    }
+    resultsForTestAndChange
+        .putIfAbsent(
+            change['name'], () => Map<String, List<Map<String, dynamic>>>())
+        .putIfAbsent(key, () => <Map<String, dynamic>>[])
+        .add(change);
+  }
+
+  configSets = Map<String, List<String>>();
+  final byBlamelist = Map<int, Map<int, Map<String, List<ByConfigSetData>>>>();
+
+  for (final test in resultsForTestAndChange.keys) {
+    for (final results in resultsForTestAndChange[test].keys) {
+      var changes = resultsForTestAndChange[test][results];
+      while (changes.isNotEmpty) {
+        var before = MinMax();
+        for (var change in changes) {
+          before.add(hashIndex[change["previous_commit_hash"]]);
+        }
+        // Sort changes by before, take all where after < first before, repeat
+        var firstSection = changes
+            .where((change) => hashIndex[change["commit_hash"]] < before.min)
+            .toList();
```

Comment:
Okay, so these are all the changes that happened after the latest preceding commit (in lack of a better term).